File: asd-millennio-frontend/lib/api/nft.ts
import { apiClient } from "./client";

export interface AcquistoNFTResponse {
  id: string;
  stripe_session_id: string;
  stripe_checkout_url: string;
  importo_eur: number;
  slot_count: number;
}

export interface DashboardFundraising {
  totale_raccolto_eur: number;
  obiettivo_eur: number;
  percentuale: number;
  nft_emessi_totali: number;
  nft_per_fascia: Record<string, number>;
}

export async function avviaAcquistoNFT(payload: {
  selezione: { slot_id: string; ore: number[] }[];
  acquisto_per_minore?: boolean;
  minore_id?: string;
}): Promise<AcquistoNFTResponse> {
  const { data } = await apiClient.post<AcquistoNFTResponse>("/nft/acquisto", payload);
  return data;
}

export async function fetchDashboardFundraising(): Promise<DashboardFundraising> {
  const { data } = await apiClient.get<DashboardFundraising>("/nft/dashboard");
  return data;
}

export async function verificaAccesso(tokenId: number, slotKey: string) {
  const { data } = await apiClient.get("/nft/verify", {
    params: { token_id: tokenId, slot_key: slotKey },
  });
  return data;
}

export interface MioNFT {
  id: string;
  token_id: number | null;
  contract_address: string | null;
  mint_tx_hash: string | null;
  ipfs_uri: string | null;
  importo_eur: number;
  ore_totali: number;
  stato: string;
  data_primo_slot: string | null;
  polygonscan_url: string | null;
  certificato_disponibile: boolean;
  created_at: string;
}

export async function fetchMieiNFT(): Promise<MioNFT[]> {
  const { data } = await apiClient.get<MioNFT[]>("/nft/le-mie");
  return data;
}

// Scarica il certificato PDF: l'endpoint richiede il Bearer token, quindi non si può
// usare un semplice <a href>. Si recupera come blob e si forza il download.
export async function scaricaCertificato(id: string): Promise<void> {
  const res = await apiClient.get(`/nft/${id}/certificato`, { responseType: "blob" });
  const url = window.URL.createObjectURL(res.data as Blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = `certificato-palasirio-${id}.pdf`;
  document.body.appendChild(a);
  a.click();
  a.remove();
  window.URL.revokeObjectURL(url);
}

File: asd-millennio-backend/src/schemas/nft.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class OreSelezioneNFT(BaseModel):
    slot_id: UUID
    ore: list[int]


class AcquistoNFTRequest(BaseModel):
    selezione: list[OreSelezioneNFT]
    acquisto_per_minore: bool = False
    minore_id: UUID | None = None


class AcquistoNFTResponse(BaseModel):
    id: UUID
    stripe_session_id: str
    stripe_checkout_url: str
    importo_eur: float
    slot_count: int


class AcquistoNFTDetail(BaseModel):
    id: UUID
    token_id: int | None
    ipfs_uri: str | None
    importo_eur: float
    stato: str
    wallet_address: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class MioNFTResponse(BaseModel):
    """Voce della pagina soci "I miei NFT"."""
    id: UUID
    token_id: int | None
    contract_address: str | None
    mint_tx_hash: str | None
    ipfs_uri: str | None
    importo_eur: float
    ore_totali: int
    stato: str
    data_primo_slot: str | None
    polygonscan_url: str | None
    certificato_disponibile: bool
    created_at: datetime


class NFTVerificaResponse(BaseModel):
    valid: bool
    socio: str | None
    slot: dict | None
    checked_at: datetime


class DashboardFundraisingResponse(BaseModel):
    totale_raccolto_eur: float
    obiettivo_eur: float
    percentuale: float
    nft_emessi_totali: int
    nft_per_fascia: dict[str, int]
